### DISCUSSION TITLE
Is there a way to query usage?

### DISCUSSION THREAD

**Discussion by ivanp** [supabase/supabase#1648] [Questions]
Is it possible to query space used and left for:

- Database 
- Users
- Storage

just like in the dashboard settings?

**Comment by awalias** [comment]
we don't yet expose this via API but it's on the roadmap

it would be possible to hack this together yourself by querying the database directly however (postgres connection string can be found in Settings > Database on the dashboard), then you can do things like `select count(id) from auth.users;`

**Reply by awalias** [reply]
actually for disk space you can define your own postgres function and call it from supabase-js:

first enable the `plpython3u` extension in Database > Extensions page in the dashboard

![image](https://user-images.githubusercontent.com/458736/119102260-67ff5680-ba4c-11eb-833b-6e757d184b0d.png)

then in SQL editor create the function:

```sql
CREATE OR REPLACE FUNCTION get_tablespace_disk_usage()
    RETURNS TABLE (
        path VARCHAR,
        bytes_free BIGINT,
        total_bytes BIGINT
    )
AS $$
import os

data_directory = plpy.execute("select setting from pg_settings where name='data_directory';")[0]['setting']
records = []

for t in plpy.execute("select spcname, spcacl, pg_tablespace_location(oid) as path from pg_tablespace"):
    if t['spcacl']:
        # TODO handle ACLs. For now only show public tablespaces.
        continue

    name = t['spcname']
    if name == 'pg_default':
        path = os.path.join(data_directory, 'default')
    elif name == 'pg_global':
        path = os.path.join(data_directory, 'global')
    else:
        path = t['path']

    # not all tablespaces actually seem to exist(?) in particular, pg_default.
    if os.path.exists(path):
        s = os.statvfs(path)
        total_bytes = s.f_blocks * s.f_frsize
        bytes_free = s.f_bavail * s.f_frsize

        records.append((path, bytes_free, total_bytes))

return records

$$ LANGUAGE plpython3u STABLE SECURITY DEFINER;
```

and query it with:
```sql
SELECT path, bytes_free, total_bytes FROM get_tablespace_disk_usage();
```

or via Javascript:

```javascript
let { data, error } = await supabase
  .rpc('get_tablespace_disk_usage')
```
